metal/eip_controlplane_reconciliation.go

```diff
@@ -267,106 +287,273 @@ func (m *controlPlaneEndpointManager) reassign(ctx context.Context, nodes []*v1.
+		return tryFilterSelf(self, nodes)
+	}
+
+	if hasIP || (len(controlPlaneEndpoint.Assignments) == 0) {
+		klog.Info("trying to reassign EIP to another node")
+		return m.tryReassign(ctx, controlPlaneEndpoint, filterDeletingNodes, tryFilterUnschedulableNodes, selfFilter)
+	}
+
+	return nil
+}
+
+// Anything calling this function should be wrapped by a lock on m.assignmentMutex
+func (m *controlPlaneEndpointManager) tryReassign(ctx context.Context, controlPlaneEndpoint *packngo.IPAddressReservation, filters ...nodeFitler) error {
+	controlPlaneHealthURL := m.healthURLFromControlPlaneEndpoint(controlPlaneEndpoint)
+	var controlPlaneNodes []*v1.Node
+
+	for _, s := range m.controlPlaneSelectors {
+		klog.V(5).Infof("tryReassign(): listing nodes with labelselector %s", s.String())
+
+		nodes, err := m.k8sclient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: s.String()})
 		if err != nil {
-			klog.V(2).Infof("failed to get endpoints %s: %v", svc.Name, err)
-			return fmt.Errorf("failed to get endpoints %s: %v", svc.Name, err)
+			return fmt.Errorf("failed to list control plane nodes with labelselector %s: %w", s.String(), err)
 		}
-		// two options:
-		// - our endpoints already exists: just copy the endpoints
-		// - our endpoints does not exist: create it
-		epExisted := true
-		myeps := m.k8sclient.CoreV1().Endpoints(externalServiceNamespace)
-		myep, err := myeps.Get(ctx, externalServiceName, metav1.GetOptions{})
-		if err != nil {
-			klog.Infof("endpoint %s/%s did not yet exist, creating", externalServiceNamespace, externalServiceName)
-			myep = &v1.Endpoints{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      externalServiceName,
-					Namespace: externalServiceNamespace,
-				},
-			}
-			epExisted = false
+
+		// TODO: dedup theese
+		for i := range nodes.Items {
+			controlPlaneNodes = append(controlPlaneNodes, &nodes.Items[i])
 		}
+	}
 
-		myep.Subsets = []v1.EndpointSubset{}
-		for _, s := range ep.Subsets {
-			copiedSubset := s.DeepCopy()
-			myep.Subsets = append(myep.Subsets, *copiedSubset)
+	potentialNodes := controlPlaneNodes
+
+	for _, f := range filters {
+		potentialNodes = f(controlPlaneNodes)
+	}
+
+	if err := m.reassign(ctx, potentialNodes, controlPlaneEndpoint, controlPlaneHealthURL); err != nil {
+		return fmt.Errorf("failed to assign the control plane endpoint: %w", err)
+	}
+
+	return nil
+}
+
+func (m *controlPlaneEndpointManager) healthURLFromControlPlaneEndpoint(controlPlaneEndpoint *packngo.IPAddressReservation) string {
+	return fmt.Sprintf("https://%s:%d/healthz", controlPlaneEndpoint.Address, m.apiServerPort)
+}
+
+func (m *controlPlaneEndpointManager) syncEndpoints(ctx context.Context, k8sEndpoints *v1.Endpoints) error {
+	m.endpointsMutex.Lock()
+	defer m.endpointsMutex.Unlock()
+
+	eps, err := m.k8sclient.CoreV1().Endpoints(externalServiceNamespace).Get(ctx, externalServiceName, metav1.GetOptions{})
+
+	switch {
+	case apierrors.IsNotFound(err):
+		externalEndpoints := &v1.Endpoints{
```

Comment:
The code in this case and the `default` case is almost identical. Would it work if we did it:

```go
	var externalEndpoints *v1.Endpoints
	switch {
	case apierrors.IsNotFound(err):
		externalEndpoints = &v1.Endpoints{
			ObjectMeta: metav1.ObjectMeta{
				Name:      externalServiceName,
				Namespace: externalServiceNamespace,
			},
		}

	case err != nil:
		return fmt.Errorf("failed to get endpoint %s/%s: %w", externalServiceNamespace, externalServiceName, err)
	default:
		externalEndpoints = eps.DeepCopy()
	}

	externalEndpoints.Subsets = make([]v1.EndpointSubset, 0, len(k8sEndpoints.Subsets))

	for _, subset := range k8sEndpoints.Subsets {
		externalEndpoints.Subsets = append(externalEndpoints.Subsets, *subset.DeepCopy())
	}

	// TODO: replace with Apply
	_, err := m.k8sclient.CoreV1().Endpoints(externalServiceNamespace).Update(ctx, externalEndpoints, metav1.UpdateOptions{})
	if err != nil {
		return fmt.Errorf("failed to update endpoint %s/%s: %w", externalServiceNamespace, externalServiceName, err)
	}

	return nil
```

Reply:
I went ahead and refactored things to use ServerSideApply to avoid the duplication.

Reply:
And I rather like serversideapply. :-)